### ISSUE TITLE
bun build injects env into bundle when --target=bun

### ISSUE THREAD

**Issue by jackson-bean** [oven-sh/bun#3835]
### What version of Bun is running?

0.7.1

### What platform is your computer?

Darwin 21.4.0 arm64 arm

### What steps can reproduce the bug?

- `bun init`

```dotenv
# .env
TEST_VALUE="hello, world!"
```
```typescript
// index.ts
console.log(process.env.TEST_VALUE);
```

- `bun build --target=bun --outdir=dist index.ts`

```javascript
// dist/index.js
console.log("hello, world!");
```

### What is the expected behavior?

Unless this is intentional, the output of the bundle should still contain "process.env.TEST_VALUE" as opposed to the value present at bundle time.

### What do you see instead?

Bun injects the .env values present at bundle time into the bundle as hardcoded values.

### Additional information

I think having the option for this unintentional behavior, potentially via a flag, could actually be beneficial as a feature. I'm considering frontend applications where while the concept of environment variables doesn't exist, a level of bundle-time configuration could be useful. Similar behavior can be found in [create-react-app](https://create-react-app.dev/docs/adding-custom-environment-variables/) and [vite](https://vitejs.dev/guide/env-and-mode.html#production-replacement).

**Comment by v3nntrnx** [comment]
oh, that's weird. doesn't happen with "Bun.env.TEST_VALUE"

**Comment by v3nntrnx** [comment]
also doesn't happen with

const test = process.env;
const test2 = test.TEST;
console.log(test2);

**Comment by pepper-lees** [comment]
> I think having the option for this unintentional behavior, potentially via a flag, could actually be beneficial as a feature. I'm considering frontend applications where while the concept of environment variables doesn't exist, a level of bundle-time configuration could be useful. Similar behavior can be found in [create-react-app](https://create-react-app.dev/docs/adding-custom-environment-variables/) and [vite](https://vitejs.dev/guide/env-and-mode.html#production-replacement).

I agree. I think it shouldn't be the default, but if replacing `process.env.<x>` with a hard-coded string does improve performance slightly (which I assume is the reason it was done in the first place), then sure, add the option to keep it enabled - but when this is done by default, it seems to break various apps that assume `process.env.<x>` are writable.

**Comment by v3nntrnx** [comment]
perhaps a env-expose option in bunfig.toml? to specify which environment variables can be "hard-coded"

**Comment by pepper-lees** [comment]
> perhaps a env-expose option in bunfig.toml? to specify which environment variables can be "hard-coded"

hmm, I would say that's a bit more complicated than is necessary. An all-or-nothing approach makes more sense to me. If someone is building a project from scratch specifically for Bun, I think it should maybe be discouraged to set process.env variables at runtime, considering the design decisions made in Bun's development and the way it "prefers" to replace `process.env.<x>` with a hardcoded string. But if someone is porting an existing Node project to bun (like what I'm doing) or is importing a library that assumes `process.env.NODE_ENV`, etc. are writable, then it would be simpler IMO to just entirely switch off the part of the preprocessor that replaces `process.env.<x>` with the appropriate strings.

**Comment by v3nntrnx** [comment]
with my option i mean you can enter:
false (default), true or an array of env which can be exposed.
i think it might be useful for some use cases like vite does (https://vitejs.dev/guide/env-and-mode.html#production-replacement)

**Comment by SEAPUNK** [comment]
I don't understand why this is implemented in the first place - environment variables tend to be the place where secrets are provided (regardless of whether they should be a way to pass down secrets in the first place).

I could see the bundler inlining environment variables that hold a specific prefix (if we're to follow [Next.js's example](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser), `process.env.PUBLIC_`), but aside from that, isn't the point of environment variables to, well, be variable for the environment? Inlining them makes it hard to do that. I'd at least expect there to be a toggle for this.

My friend got bit by this a few minutes ago, as they were bundling their app to deploy somewhere, but it was inlining their database login information into the bundle (since they had a `.env` file locally), despite them wanting to explicitly provide it through environment variables at runtime.

**Comment by keverw** [comment]
I ran into this issue, and thought I wasn't passing the ENVs correctly. I agree this should be configurable, since I thought I was going crazy at first until I dug into it and figured out this is what was happen.

For now, `globalThis.process.env.NODE_ENV` works as a work around for anyone else needing to access the real actual ENV values.

**Comment by tanishqmanuja** [comment]
```
import process from "process";
```
Also fixes this for now.

**Comment by atbe** [comment]
> ```
> import process from "process";
> ```
> 
> Also fixes this for now.

Can confirm this works -- @Jarred-Sumner this needs to be more explicit in the documentation because environment variable values are being embedded in the generated bundles/binaries, causing `process.env.SOME_ENV_VAR` to be replaced with strings that reflect the value at build-time, locking the string in place and preventing any form of overriding via environment variables at runtime.

**Comment by ChemaCLi** [comment]
It would be very nice to have a  `--no-env-injection` param or a way to setup the bun behaviours in a bunconfig.json or something like that

**Comment by rubenfiszel** [comment]
Any updates on this? This is affecting our production builds, we have to add:
`import process from "process";` everywhere

**Comment by paperdave** [comment]
i swear i had done this, but #8576 is closed. i think the solution i wanted was to have `--env-prefix` to allow exposing *some* environment variables. but regardless, this is not supposed to be enabled.

**Comment by kobutri** [comment]
Imho inlining variables is a nice feature, but not a good default. I think something like a `Bun.build.env` that inlines the variables at build time would be nice.

**Comment by WiLuX-Source** [comment]
I'm also experiencing this issue I'll try importing process. Does targeting bun or node have any difference in terms of performance?

**Comment by gbarton** [comment]
Ran into this today when building.  Please make this a flag or something to disable during build.  The documentation states that one does not need dotenv because bun reads .env files fine.  It clearly does not when you attempt to deploy.

Current workaround of using `Bun.env.VAR` instead of `process.env.VAR` does work.

**Comment by RiskyMH** [comment]
We now have [`--env=disable`](https://bun.sh/docs/bundler#env) as of [`v1.1.39`](https://bun.sh/blog/bun-v1.1.39#inject-environment-variables-in-bun-build). Also as of `1.2` it shouldn't inline by default.

**Comment by kelvinwop** [comment]
<img width="813" height="671" alt="Image" src="https://github.com/user-attachments/assets/3bcd39ea-d792-4fe8-b696-bd2b658062b6" />

does not transform Bun.env.XXXX, only process.env.XXXX
